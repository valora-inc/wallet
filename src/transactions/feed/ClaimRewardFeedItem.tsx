import BigNumber from 'bignumber.js'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import AppAnalytics from 'src/analytics/AppAnalytics'
import { HomeEvents } from 'src/analytics/Events'
import TokenDisplay from 'src/components/TokenDisplay'
import Touchable from 'src/components/Touchable'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import Colors from 'src/styles/colors'
import { typeScale } from 'src/styles/fonts'
import { Spacing } from 'src/styles/styles'
import variables from 'src/styles/variables'
import TransactionFeedItemImage from 'src/transactions/feed/TransactionFeedItemImage'
import { ClaimReward } from 'src/transactions/types'

interface DescriptionProps {
  transaction: ClaimReward
}

function Description({ transaction }: DescriptionProps) {
  const { t } = useTranslation()
  const txAppName = transaction.appName
  const title = t('transactionFeed.claimRewardTitle')
  const subtitle = t('transactionFeed.claimRewardSubtitle', {
    context: !txAppName ? 'noTxAppName' : undefined,
    txAppName,
  })

  return (
    <View style={styles.contentContainer}>
      <Text style={styles.title} testID={'ClaimRewardFeedItem/title'} numberOfLines={1}>
        {title}
      </Text>
      <Text style={styles.subtitle} testID={'ClaimRewardFeedItem/subtitle'} numberOfLines={1}>
        {subtitle}
      </Text>
    </View>
  )
}

interface AmountDisplayProps {
  transaction: ClaimReward
  isLocal: boolean
}

function AmountDisplay({ transaction, isLocal }: AmountDisplayProps) {
  const amountValue = new BigNumber(transaction.amount.value)
  const tokenId = transaction.amount.tokenId

  const textStyle = isLocal ? styles.amountSubtitle : [styles.amountTitle, { color: Colors.accent }]

  return (
    <TokenDisplay
      amount={amountValue}
      localAmount={transaction.amount.localAmount}
      tokenId={tokenId}
      showLocalAmount={isLocal}
      showSymbol={true}
      showExplicitPositiveSign={!isLocal}
      hideSign={!!isLocal}
      style={textStyle}
      testID={`ClaimRewardFeedItem/amount-${isLocal ? 'local' : 'crypto'}`}
    />
  )
}

interface AmountProps {
  transaction: ClaimReward
}

function Amount({ transaction }: AmountProps) {
  return (
    <View style={styles.amountContainer}>
      <AmountDisplay transaction={transaction} isLocal={false} />
      <AmountDisplay transaction={transaction} isLocal={true} />
    </View>
  )
}

interface Props {
  transaction: ClaimReward
}

export default function ClaimRewardFeedItem({ transaction }: Props) {
  return (
    <Touchable
      testID={`ClaimRewardFeedItem/${transaction.transactionHash}`}
      onPress={() => {
        AppAnalytics.track(HomeEvents.transaction_feed_item_select, {
          itemType: transaction.type,
        })
        navigate(Screens.TransactionDetailsScreen, { transaction })
      }}
    >
      <View style={styles.container}>
        <TransactionFeedItemImage
          status={transaction.status}
          transactionType={transaction.type}
          networkId={transaction.networkId}
        />
        <Description transaction={transaction} />
        <Amount transaction={transaction} />
      </View>
    </Touchable>
  )
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    paddingVertical: Spacing.Small12,
    paddingHorizontal: variables.contentPadding,
    alignItems: 'center',
  },
  contentContainer: {
    flex: 1,
    paddingHorizontal: variables.contentPadding,
  },
  title: {
    ...typeScale.labelMedium,
    color: Colors.black,
  },
  subtitle: {
    ...typeScale.bodySmall,
    color: Colors.gray4,
  },
  amountContainer: {
    maxWidth: '50%',
  },
  amountTitle: {
    ...typeScale.labelMedium,
    color: Colors.black,
    flexWrap: 'wrap',
    textAlign: 'right',
  },
  amountSubtitle: {
    ...typeScale.bodySmall,
    color: Colors.gray4,
    flexWrap: 'wrap',
    textAlign: 'right',
  },
})
