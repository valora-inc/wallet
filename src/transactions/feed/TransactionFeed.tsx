import React, { useMemo } from 'react'
import { ActivityIndicator, SectionList, StyleSheet, View } from 'react-native'
import SectionHead from 'src/components/SectionHead'
import useSelector from 'src/redux/useSelector'
import colors from 'src/styles/colors'
import { Spacing } from 'src/styles/styles'
import NoActivity from 'src/transactions/NoActivity'
import NftFeedItem from 'src/transactions/feed/NftFeedItem'
import SwapFeedItem from 'src/transactions/feed/SwapFeedItem'
import TransferFeedItem from 'src/transactions/feed/TransferFeedItem'
import {
  deduplicateTransactions,
  getAllowedNetworkIds,
  useFetchTransactions,
} from 'src/transactions/feed/queryHelper'
import {
  completedStandbyTransactionsSelector,
  pendingStandbyTransactionsSelector,
  transactionsSelector,
} from 'src/transactions/reducer'
import { TokenTransaction, TransactionStatus } from 'src/transactions/types'
import { groupFeedItemsInSections } from 'src/transactions/utils'

export type FeedTokenProperties = {
  status: TransactionStatus // for standby transactions
}

export type FeedTokenTransaction = TokenTransaction & FeedTokenProperties

function TransactionFeed() {
  const { loading, error, transactions, fetchingMoreTransactions, fetchMoreTransactions } =
    useFetchTransactions()

  const cachedTransactions = useSelector(transactionsSelector)
  const allPendingTransactions = useSelector(pendingStandbyTransactionsSelector)
  const allConfirmedStandbyTransactions = useSelector(completedStandbyTransactionsSelector)
  const allowedNetworks = getAllowedNetworkIds()

  const confirmedFeedTransactions = useMemo(() => {
    const confirmedTokenTransactions: TokenTransaction[] =
      transactions.length > 0 ? transactions : cachedTransactions
    const allConfirmedTransactions = deduplicateTransactions(
      confirmedTokenTransactions,
      allConfirmedStandbyTransactions
    )
    return allConfirmedTransactions.filter((tx) => {
      return allowedNetworks.includes(tx.networkId)
    })
  }, [transactions, cachedTransactions, allowedNetworks, allConfirmedStandbyTransactions])

  const pendingTransactions = useMemo(() => {
    return allPendingTransactions.filter((tx) => {
      return allowedNetworks.includes(tx.networkId)
    })
  }, [allPendingTransactions, allowedNetworks])

  const sections = useMemo(() => {
    if (confirmedFeedTransactions.length === 0 && pendingTransactions.length === 0) {
      return []
    }

    return groupFeedItemsInSections(pendingTransactions, confirmedFeedTransactions)
  }, [pendingTransactions, confirmedFeedTransactions])

  if (!sections.length) {
    return <NoActivity loading={loading} error={error} />
  }

  function renderItem({ item: tx }: { item: FeedTokenTransaction; index: number }) {
    switch (tx.__typename) {
      case 'TokenExchangeV3':
        return <SwapFeedItem key={tx.transactionHash} exchange={tx} />
      case 'TokenTransferV3':
        return <TransferFeedItem key={tx.transactionHash} transfer={tx} />
      case 'NftTransferV3':
        return <NftFeedItem key={tx.transactionHash} transaction={tx} />
    }
  }

  return (
    <>
      <SectionList
        renderItem={renderItem}
        renderSectionHeader={(item) => <SectionHead text={item.section.title} />}
        sections={sections}
        keyExtractor={(item) => `${item.transactionHash}-${item.timestamp.toString()}`}
        keyboardShouldPersistTaps="always"
        testID="TransactionList"
        onEndReached={() => fetchMoreTransactions()}
      />
      {fetchingMoreTransactions && (
        <View style={styles.centerContainer}>
          <ActivityIndicator style={styles.loadingIcon} size="large" color={colors.greenBrand} />
        </View>
      )}
    </>
  )
}

const styles = StyleSheet.create({
  loadingIcon: {
    marginVertical: Spacing.Thick24,
    height: 108,
    width: 108,
  },
  centerContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
})

export default TransactionFeed
