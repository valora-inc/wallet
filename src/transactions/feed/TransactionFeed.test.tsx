import { fireEvent, render, waitFor, within } from '@testing-library/react-native'
import { FetchMock } from 'jest-fetch-mock/types'
import React from 'react'
import { Provider } from 'react-redux'
import { ReactTestInstance } from 'react-test-renderer'
import { RootState } from 'src/redux/reducers'
import { getDynamicConfigParams, getFeatureGate, getMultichainFeatures } from 'src/statsig'
import { StatsigFeatureGates } from 'src/statsig/types'
import { QueryResponse } from 'src/transactions/feed/queryHelper'
import TransactionFeed from 'src/transactions/feed/TransactionFeed'
import {
  NetworkId,
  StandbyTransaction,
  TokenTransaction,
  TokenTransactionTypeV2,
  TransactionStatus,
} from 'src/transactions/types'
import networkConfig from 'src/web3/networkConfig'
import { createMockStore, RecursivePartial } from 'test/utils'
import { mockApprovalTransaction, mockCusdAddress, mockCusdTokenId } from 'test/values'

jest.mock('src/statsig')

const mockTransaction = (
  transactionHash: string,
  status = TransactionStatus.Complete
): TokenTransaction => {
  return {
    networkId: NetworkId['celo-alfajores'],
    address: '0xd68360cce1f1ff696d898f58f03e0f1252f2ea33',
    amount: {
      tokenId: mockCusdTokenId,
      tokenAddress: mockCusdAddress,
      value: '0.1',
    },
    block: '8648978',
    fees: [],
    metadata: {},
    timestamp: 1542306118,
    transactionHash,
    type: TokenTransactionTypeV2.Received,
    status,
  }
}

const STAND_BY_TRANSACTION_SUBTITLE_KEY = 'confirmingTransaction'

const MOCK_STANDBY_TRANSACTION: StandbyTransaction = {
  context: { id: 'test' },
  networkId: NetworkId['celo-alfajores'],
  type: TokenTransactionTypeV2.Sent,
  status: TransactionStatus.Pending,
  amount: {
    value: '0.5',
    tokenAddress: mockCusdAddress,
    tokenId: mockCusdTokenId,
  },
  metadata: {},
  timestamp: 1542300000,
  address: '0xd68360cce1f1ff696d898f58f03e0f1252f2ea33',
}

const END_CURSOR = 'YXJyYXljb25uZWN0aW9uOjk='

const MOCK_EMPTY_RESPONSE: QueryResponse = {
  data: {
    tokenTransactionsV3: {
      pageInfo: {
        startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
        endCursor: END_CURSOR,
        hasNextPage: true,
        hasPreviousPage: false,
      },
      transactions: [],
    },
  },
}

const MOCK_EMPTY_RESPONSE_NO_NEXT_PAGE: QueryResponse = {
  data: {
    tokenTransactionsV3: {
      pageInfo: {
        startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
        endCursor: END_CURSOR,
        hasNextPage: false,
        hasPreviousPage: false,
      },
      transactions: [],
    },
  },
}

const MOCK_RESPONSE: QueryResponse = {
  data: {
    tokenTransactionsV3: {
      pageInfo: {
        startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
        endCursor: END_CURSOR,
        hasNextPage: true,
        hasPreviousPage: false,
      },
      transactions: [
        mockTransaction('0x544367eaf2b01622dd1c7b75a6b19bf278d72127aecfb2e5106424c40c268e8b2'),
      ],
    },
  },
}

const MOCK_RESPONSE_NO_NEXT_PAGE: QueryResponse = {
  data: {
    tokenTransactionsV3: {
      pageInfo: {
        startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
        endCursor: END_CURSOR,
        hasNextPage: false,
        hasPreviousPage: false,
      },
      transactions: [
        mockTransaction('0x544367eaf2b01622dd1c7b75a6b19bf278d72127aecfb2e5106424c40c268e8b'),
      ],
    },
  },
}

const MOCK_RESPONSE_MANY_ITEMS: QueryResponse = {
  data: {
    tokenTransactionsV3: {
      pageInfo: {
        startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
        endCursor: END_CURSOR,
        hasNextPage: true,
        hasPreviousPage: false,
      },
      transactions: [...Array(10).keys()].map((id) => mockTransaction(id.toString())),
    },
  },
}

const MOCK_RESPONSE_FAILED_TRANSACTION: QueryResponse = {
  data: {
    tokenTransactionsV3: {
      pageInfo: {
        startCursor: 'YXJyYXljb25uZWN0aW9uOjA=',
        endCursor: END_CURSOR,
        hasNextPage: false,
        hasPreviousPage: false,
      },
      transactions: [
        mockTransaction(
          '0x544367eaf2b01622dd1c7b75a6b19bf278d72127aecfb2e5106424c40c268e8b',
          TransactionStatus.Failed
        ),
      ],
    },
  },
}

describe('TransactionFeed', () => {
  const mockFetch = fetch as FetchMock
  beforeEach(() => {
    jest.clearAllMocks()
    jest.mocked(getMultichainFeatures).mockReturnValue({
      showCico: [NetworkId['celo-alfajores']],
      showBalances: [NetworkId['celo-alfajores']],
      showTransfers: [NetworkId['celo-alfajores']],
      showApprovalTxsInHomefeed: [NetworkId['celo-alfajores']],
    })
    jest.mocked(getDynamicConfigParams).mockReturnValue({
      jumpstartContracts: {
        ['celo-alfajores']: { contractAddress: '0x7bf3fefe9881127553d23a8cd225a2c2442c438c' },
      },
    })
    mockFetch.resetMocks()
  })

  function renderScreen(storeOverrides: RecursivePartial<RootState> = {}) {
    const store = createMockStore({
      ...storeOverrides,
    })

    const tree = render(
      <Provider store={store}>
        <TransactionFeed />
      </Provider>
    )

    return {
      store,
      ...tree,
    }
  }

  function getNumTransactionItems(sectionList: ReactTestInstance) {
    // data[0] is the first section in the section list - all mock transactions
    // are for the same section / date
    return sectionList.props.data[0].data.length
  }

  it('only renders approval txs from supported networks', async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_EMPTY_RESPONSE_NO_NEXT_PAGE))

    const tree = renderScreen({
      transactions: {
        transactionsByNetworkId: {
          [NetworkId['ethereum-sepolia']]: [mockApprovalTransaction],
          [NetworkId['celo-alfajores']]: [
            {
              ...mockApprovalTransaction,
              networkId: NetworkId['celo-alfajores'],
              transactionHash: '0xfoo',
            },
          ],
        },
        standbyTransactions: [],
      },
    })
    await waitFor(() => expect(tree.getByTestId('TransactionList').props.data.length).toBe(1))

    expect(tree.queryByTestId('NoActivity/loading')).toBeNull()
    expect(tree.queryByTestId('NoActivity/error')).toBeNull()
    expect(mockFetch).toHaveBeenCalledTimes(1)
    expect(tree.getAllByTestId(new RegExp('TokenApprovalFeedItem', 'i')).length).toBe(1)
    expect(tree.queryByTestId(`TokenApprovalFeedItem/0xfoo`)).not.toBeNull()
  })

  it('renders correctly when there is a response', async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE))

    const tree = renderScreen({})
    await waitFor(() => expect(tree.getByTestId('TransactionList').props.data.length).toBe(1))

    expect(tree.queryByTestId('NoActivity/loading')).toBeNull()
    expect(tree.queryByTestId('NoActivity/error')).toBeNull()
    expect(mockFetch).toHaveBeenCalledTimes(1)
    expect(tree.getAllByTestId('TransferFeedItem').length).toBe(1)
    expect(
      within(tree.getByTestId('TransferFeedItem')).getByTestId('TransferFeedItem/title')
    ).toHaveTextContent(
      'feedItemReceivedTitle, {"displayName":"feedItemAddress, {\\"address\\":\\"0xd683...ea33\\"}"}'
    )
  })

  it('renders correctly with completed standby transactions', async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE))

    const tree = renderScreen({
      transactions: {
        standbyTransactions: [
          {
            ...MOCK_STANDBY_TRANSACTION,
            status: TransactionStatus.Complete,
            transactionHash: '0x544367eaf2b01622dd1c7b75a6b19bf278d72127aecfb2e5106424c40c268e8bxx',
            block: '8888',
            fees: [],
          },
        ],
      },
    })

    await waitFor(() => expect(tree.getAllByTestId('TransferFeedItem').length).toBe(2))
  })

  it("doesn't render transfers for tokens that we don't know about", async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE))

    const { getAllByTestId, getByTestId } = renderScreen({})

    await waitFor(() => getByTestId('TransactionList'))

    const items = getAllByTestId('TransferFeedItem/title')
    expect(items.length).toBe(1)
  })

  it('renders the loading indicator while it loads', async () => {
    const { getByTestId, queryByTestId } = renderScreen({})
    expect(getByTestId('NoActivity/loading')).toBeDefined()
    expect(queryByTestId('NoActivity/error')).toBeNull()
    expect(queryByTestId('TransactionList')).toBeNull()
  })

  it("renders an error screen if there's no cache and the query fails", async () => {
    mockFetch.mockReject(new Error('Test error'))

    const { getByTestId, queryByTestId } = renderScreen({})
    await waitFor(() => getByTestId('NoActivity/error'))
    expect(queryByTestId('NoActivity/loading')).toBeNull()
    expect(queryByTestId('TransactionList')).toBeNull()
  })

  it('renders the cache if there is one', async () => {
    mockFetch.mockReject(new Error('Test error'))

    const { getByTestId, queryByTestId } = renderScreen({
      transactions: {
        transactionsByNetworkId: {
          [networkConfig.defaultNetworkId]: MOCK_RESPONSE.data.tokenTransactionsV3.transactions,
        },
      },
    })

    expect(queryByTestId('NoActivity/loading')).toBeNull()
    expect(queryByTestId('NoActivity/error')).toBeNull()
    expect(getByTestId('TransactionList')).not.toBeNull()
  })

  it('renders correctly when there are confirmed transactions and stand by transactions', async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE))

    const tree = renderScreen({
      transactions: {
        standbyTransactions: [MOCK_STANDBY_TRANSACTION],
      },
    })

    await waitFor(() => tree.getByTestId('TransactionList'))

    expect(tree.queryByTestId('NoActivity/loading')).toBeNull()
    expect(tree.queryByTestId('NoActivity/error')).toBeNull()

    const subtitles = tree.queryAllByTestId('TransferFeedItem/subtitle')

    const pendingSubtitles = subtitles.filter((node) =>
      node.children.some((ch) => ch === STAND_BY_TRANSACTION_SUBTITLE_KEY)
    )
    expect(pendingSubtitles.length).toBe(1)
  })

  it('renders correct status for a complete transaction', async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE))

    const { getByTestId, getByText } = renderScreen({})

    await waitFor(() => getByTestId('TransactionList'))

    expect(getByText('feedItemReceivedInfo, {"context":"noComment"}')).toBeTruthy()
  })

  it('renders correct status for a failed transaction', async () => {
    mockFetch.mockResponse(JSON.stringify(MOCK_RESPONSE_FAILED_TRANSACTION))

    const { getByTestId, getByText } = renderScreen({})

    await waitFor(() => getByTestId('TransactionList'))

    expect(getByText('feedItemFailedTransaction')).toBeTruthy()
  })

  it('tries to fetch 10 transactions, unless the end is reached', async () => {
    mockFetch.mockImplementation((url: any, request: any) => {
      const body: string = request.body
      let response = ''
      if (body.includes(END_CURSOR)) {
        response = JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE)
      } else {
        response = JSON.stringify(MOCK_RESPONSE)
      }
      return Promise.resolve(new Response(response))
    })

    const tree = renderScreen({})

    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(2))
    expect(getNumTransactionItems(tree.getByTestId('TransactionList'))).toBe(2)
  })

  it('tries to fetch 10 transactions, and stores empty pages', async () => {
    mockFetch.mockImplementation((url: any, request: any) => {
      const body: string = request.body
      let response = ''
      if (body.includes(END_CURSOR)) {
        response = JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE)
      } else {
        response = JSON.stringify(MOCK_EMPTY_RESPONSE)
      }
      return Promise.resolve(new Response(response))
    })

    const tree = renderScreen({})

    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(2))
    expect(getNumTransactionItems(tree.getByTestId('TransactionList'))).toBe(1)
  })

  it('fetches the next page by scrolling to the end of the list', async () => {
    mockFetch.mockImplementation((url: any, request: any) => {
      const body: string = request.body
      let response = ''
      if (body.includes(END_CURSOR)) {
        response = JSON.stringify(MOCK_RESPONSE_NO_NEXT_PAGE)
      } else {
        response = JSON.stringify(MOCK_RESPONSE_MANY_ITEMS)
      }
      return Promise.resolve(new Response(response))
    })

    const tree = renderScreen({})
    await waitFor(() => tree.getByTestId('TransactionList'))

    expect(mockFetch).toHaveBeenCalledTimes(1)
    expect(getNumTransactionItems(tree.getByTestId('TransactionList'))).toBe(10)

    fireEvent(tree.getByTestId('TransactionList'), 'onEndReached')
    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(2))
    expect(getNumTransactionItems(tree.getByTestId('TransactionList'))).toBe(11)
  })

  it('fetches the next page automatically if there are no transactions returned and next page exists', async () => {
    let mockFetchCount = 0
    mockFetch.mockImplementation(() => {
      let response = ''
      switch (mockFetchCount) {
        case 1:
          response = JSON.stringify(MOCK_EMPTY_RESPONSE)
          break
        case 2:
          response = JSON.stringify(MOCK_RESPONSE)
          break
        default:
          response = JSON.stringify(MOCK_RESPONSE_MANY_ITEMS)
      }
      mockFetchCount += 1
      return Promise.resolve(new Response(response))
    })

    const tree = renderScreen({})
    await waitFor(() => tree.getByTestId('TransactionList'))

    expect(mockFetch).toHaveBeenCalledTimes(1)
    expect(getNumTransactionItems(tree.getByTestId('TransactionList'))).toBe(10)

    fireEvent(tree.getByTestId('TransactionList'), 'onEndReached')
    await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(3))
    expect(getNumTransactionItems(tree.getByTestId('TransactionList'))).toBe(11)
  })

  it('renders GetStarted if SHOW_GET_STARTED is enabled and transaction feed is empty', async () => {
    jest.mocked(getFeatureGate).mockImplementation((gate) => {
      if (gate === StatsigFeatureGates.SHOW_GET_STARTED) {
        return true
      }
      if (gate === StatsigFeatureGates.SHOW_UK_COMPLIANT_VARIANT) {
        return false
      }
      throw new Error('Unexpected gate')
    })

    const { getByTestId } = renderScreen({})
    expect(getByTestId('GetStarted')).toBeDefined()
  })

  it('renders NoActivity for UK compliance', () => {
    jest.mocked(getFeatureGate).mockImplementation((gate) => {
      if (gate === StatsigFeatureGates.SHOW_GET_STARTED) {
        return true
      }
      if (gate === StatsigFeatureGates.SHOW_UK_COMPLIANT_VARIANT) {
        return true
      }
      throw new Error('Unexpected gate')
    })

    const { getByTestId, getByText } = renderScreen({})

    expect(getByTestId('NoActivity/loading')).toBeTruthy()
    expect(getByText('transactionFeed.noTransactions')).toBeTruthy()
  })

  it('renders NoActivity by default if transaction feed is empty', async () => {
    jest.mocked(getFeatureGate).mockReturnValue(false)
    const { getByTestId, getByText } = renderScreen({})
    expect(getByTestId('NoActivity/loading')).toBeDefined()
    expect(getByText('transactionFeed.noTransactions')).toBeTruthy()
  })
})
