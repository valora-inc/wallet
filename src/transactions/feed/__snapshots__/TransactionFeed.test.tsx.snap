// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionFeed renders correctly when there is a response 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "data": Array [
          Object {
            "__typename": "TokenTransferV2",
            "address": "0xd68360cce1f1ff696d898f58f03e0f1252f2ea33",
            "amount": Object {
              "tokenAddress": "0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1",
              "value": "0.1",
            },
            "block": "8648978",
            "fees": Array [],
            "metadata": Object {},
            "status": "Complete",
            "timestamp": 1542306118,
            "transactionHash": "0x544367eaf2b01622dd1c7b75a6b19bf278d72127aecfb2e5106424c40c268e8b",
            "type": "RECEIVED",
          },
        ],
        "title": "November 2018",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  keyboardShouldPersistTaps="always"
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  testID="TransactionList"
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 8,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Medium",
              "fontSize": 13,
              "lineHeight": 16,
            }
          }
        >
          November 2018
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "justifyContent": "center",
                    },
                    Object {
                      "backgroundColor": "hsl(192, 53%, 93%)",
                      "borderRadius": 20,
                      "height": 40,
                      "width": 40,
                    },
                  ]
                }
              >
                <svg
                  fill="hsl(192, 53%, 93%)"
                  height="40"
                  style={Object {}}
                  viewBox="0 0 40 40"
                  width="40"
                />
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "marginLeft": 16,
                "width": "55%",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 16,
                  "lineHeight": 22,
                }
              }
              testID="TransferFeedItem/title"
            >
              feedItemReceivedTitle, {"displayName":"feedItemAddress, {\\"address\\":\\"0xd683...ea33\\"}"}
            </Text>
            <Text
              style={
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                  "paddingTop": 2,
                }
              }
              testID="TransferFeedItem/subtitle"
            >
              feedItemReceivedInfo, {"context":"noComment"}
            </Text>
          </View>
          <View
            style={
              Object {
                "flexShrink": 1,
                "marginLeft": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 16,
                    "lineHeight": 22,
                    "marginLeft": "auto",
                    "textAlign": "right",
                  },
                  Object {
                    "color": "#1AB775",
                  },
                ]
              }
              testID="TransferFeedItem/amount"
            >
              +
              â‚±
              0.13
            </Text>
            <Text
              style={
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                  "marginLeft": "auto",
                  "minWidth": "40%",
                  "paddingTop": 2,
                  "textAlign": "right",
                }
              }
              testID="TransferFeedItem/tokenAmount"
            >
              
              0.10
               cUSD
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
  </View>
</RCTScrollView>
`;
